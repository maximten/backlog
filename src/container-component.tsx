import styled from 'styled-components';

export const ContainerComponent = styled.div`
  max-width: 768px;
  width: 100%;
  margin: auto;
`;
