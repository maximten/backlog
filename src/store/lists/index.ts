import { Action, ListsState, List, ActionType } from 'src/store/lists/types';
import { parseState, generateItemKey, generateListKey } from 'src/store/lists/parsers';
import { focusedListKeySelector, focusedListSelector } from 'src/store/lists/selectors';
import { modulo } from 'src/utils/modulo';

const updateFocusedList = (state: ListsState,
  modifier: (state: ListsState, list: List) => Partial<List>): ListsState => {
  // TODO: refactoring
  const listKey = focusedListKeySelector({ lists: state });
  const list = focusedListSelector({ lists: state });
  return {
    ...state,
    listsMap: {
      ...state.listsMap,
      [listKey]: {
        ...list,
        ...modifier(state, list),
      },
    },
  };
};

export const lists = (state: ListsState, action: Action): ListsState => {
  switch (action.type) {
    case (ActionType.SET_ITEM): {
      return updateFocusedList(state, (state, list) => ({
        items: {
          ...list.items,
          [list.focusedItem]: action.payload,
        },
      }));
    }
    case (ActionType.ADD_ITEM): {
      const itemKey = generateItemKey();
      return updateFocusedList(state, (state, list) => ({
        items: {
          ...list.items,
          [itemKey]: action.payload,
        },
        focusedItem: itemKey,
        itemsOrder: [...list.itemsOrder, itemKey],
      }));
    }
    case (ActionType.SET_FOCUSED_ITEM_KEY): {
      return updateFocusedList(state, () => ({ focusedItem: action.payload }));
    }
    case (ActionType.REMOVE_ITEM): {
      return updateFocusedList(state, (state, list) => {
        const {
          focusedItem, itemsOrder,
        } = list;
        if (itemsOrder.length === 1) {
          return;
        }
        const nextFocusedItem = itemsOrder[itemsOrder.indexOf(focusedItem) - 1];
        const items = { ...list.items };
        delete items[focusedItem];
        return {
          items,
          focusedItem: nextFocusedItem,
          itemsOrder: itemsOrder.filter((key) => key !== focusedItem),
        };
      });
    }
    case (ActionType.SWAP_ITEMS_UP): {
      return updateFocusedList(state, (state, list) => {
        const {
          focusedItem, itemsOrder,
        } = list;
        const itemIndex = itemsOrder.indexOf(focusedItem);
        const upperItemIndex = itemIndex - 1;
        if (upperItemIndex < 0) {
          return;
        }
        const upperItemKey = itemsOrder[upperItemIndex];
        const newItemsOrder = [...itemsOrder];
        newItemsOrder[itemIndex] = upperItemKey;
        newItemsOrder[upperItemIndex] = focusedItem;
        return { itemsOrder: newItemsOrder };
      });
    }
    case (ActionType.SWAP_ITEMS_DOWN): {
      return updateFocusedList(state, (state, list) => {
        const {
          focusedItem, itemsOrder,
        } = list;
        const itemIndex = itemsOrder.indexOf(focusedItem);
        const bottomItemIndex = itemIndex + 1;
        if (bottomItemIndex >= itemsOrder.length) {
          return;
        }
        const bottomItemKey = itemsOrder[bottomItemIndex];
        const newItemsOrder = [...itemsOrder];
        newItemsOrder[itemIndex] = bottomItemKey;
        newItemsOrder[bottomItemIndex] = focusedItem;
        return { itemsOrder: newItemsOrder };
      });
    }
    case (ActionType.SET_LIST_NAME): {
      return updateFocusedList(state, () => ({ name: action.payload }));
    }
    case (ActionType.ADD_LIST): {
      const listKey = generateListKey();
      return {
        ...state,
        listsMap: {
          ...state.listsMap, [listKey]: action.payload,
        },
        listsOrder: [...state.listsOrder, listKey],
        focusedList: listKey,
      };
    }
    case (ActionType.FOCUS_TO_LEFT_LIST): {
      const { listsOrder } = state;
      if (listsOrder.length === 1) {
        return;
      }
      const listKey = focusedListKeySelector({ lists: state });
      const listIndex = listsOrder.indexOf(listKey);
      const nextIndex = modulo(listIndex - 1, listsOrder.length);
      const nextKey = listsOrder[nextIndex];
      return {
        ...state,
        focusedList: nextKey,
      };
    }
    case (ActionType.FOCUS_TO_RIGHT_LIST): {
      const { listsOrder } = state;
      if (listsOrder.length === 1) {
        return;
      }
      const listKey = focusedListKeySelector({ lists: state });
      const listIndex = listsOrder.indexOf(listKey);
      const nextIndex = modulo(listIndex + 1, listsOrder.length);
      const nextKey = listsOrder[nextIndex];
      return {
        ...state,
        focusedList: nextKey,
      };
    }
    default:
      return parseState(null);
  }
};
