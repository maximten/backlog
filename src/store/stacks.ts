const DEFAULT_STACK_KEY = 0;

const EMPTY_ITEM = {
  id: 0,
  title: '',
  content: '',
};

const DEFAULT_STACKS = { [DEFAULT_STACK_KEY]: [EMPTY_ITEM] };

const DEFAULT_STACKS_INDEXES = { [DEFAULT_STACK_KEY]: 0 };

const DEFAULT_STACKS_NAMES = { [DEFAULT_STACK_KEY]: 'default' };

const DEFAULT_STACKS_ORDER = [
  DEFAULT_STACK_KEY,
];

const INITIAL_STATE = {
  stacks: DEFAULT_STACKS,
  lastKey: DEFAULT_STACK_KEY,
  order: DEFAULT_STACKS_ORDER,
  focusedStack: DEFAULT_STACK_KEY,
  focusedItem: null,
  stacksIndexes: DEFAULT_STACKS_INDEXES,
  stacksNames: DEFAULT_STACKS_NAMES,
};

const ACTIONS = {
  SET_STACKS: 'SET_STACKS',
  SET_LAST_KEY: 'SET_LAST_KEY',
  SET_ORDER: 'SET_ORDER',
  SET_FOCUSED_STACK: 'SET_FOCUSED_STACK',
  SET_FOCUSED_ITEM: 'SET_FOCUSED_ITEM',
  SET_STACK_INDEX: 'SET_STACK_INDEX',
  SET_STACK_INDEXES: 'SET_STACK_INDEXES',
  SET_STACKS_NAMES: 'SET_STACK_NAMES',
};

export const setStacks = (stacks) => ({
  type: ACTIONS.SET_STACKS,
  payload: stacks,
});

export const setLastKey = (key) => ({
  type: ACTIONS.SET_LAST_KEY,
  payload: key,
});

export const setOrder = (order) => ({
  type: ACTIONS.SET_ORDER,
  payload: order,
});

export const setFocusedStack = (key) => ({
  type: ACTIONS.SET_FOCUSED_STACK,
  payload: key,
});

export const setFocusedItem = (index) => ({
  type: ACTIONS.SET_FOCUSED_ITEM,
  payload: index,
});

export const setStackIndex = ({
  key, index,
}) => ({
  type: ACTIONS.SET_STACK_INDEX,
  payload: {
    key, index,
  },
});

export const setStackIndexes = (indexes) => ({
  type: ACTIONS.SET_STACK_INDEXES,
  payload: indexes,
});

export const setStackNames = (names) => ({
  type: ACTIONS.SET_STACKS_NAMES,
  payload: names,
});

const parseState = (data) => {
  const {
    stacks = DEFAULT_STACKS,
    lastKey = DEFAULT_STACK_KEY,
    order = DEFAULT_STACKS_ORDER,
    focusedStack = DEFAULT_STACK_KEY,
    focusedItem = null,
    stacksIndexes = DEFAULT_STACKS_INDEXES,
    stacksNames = DEFAULT_STACKS_NAMES,
  } = data;
  return {
    stacks, lastKey, order, focusedStack, focusedItem, stacksIndexes, stacksNames,
  };
};

export const fetchState = () => async (dispatch) => {
  const data = await fetch('api/stacks').then((res) => res.json());
  const {
    stacks, lastKey, order, focusedStack, focusedItem, stacksIndexes, stacksNames,
  } = parseState(data);
  dispatch(setStacks(stacks));
  dispatch(setLastKey(lastKey));
  dispatch(setOrder(order));
  dispatch(setFocusedStack(focusedStack));
  dispatch(setFocusedItem(focusedItem));
  dispatch(setStackIndexes(stacksIndexes));
  dispatch(setStackNames(stacksNames));
};

export const postState = () => async (dispatch, getState) => {
  const {
    stacks: {
      stacks,
      key,
      lastKey,
      order,
      focusedStack,
      focusedItem,
      stacksIndexes,
      stacksNames,
    },
  } = getState();
  await fetch('api/stacks', {
    method: 'POST',
    headers: { 'Content-type': 'application/json' },
    body: JSON.stringify({
      stacks,
      key,
      lastKey,
      order,
      focusedStack,
      focusedItem,
      stacksIndexes,
      stacksNames,
    }),
  });
};

export const stacks = (state = INITIAL_STATE, {
  type, payload,
}) => {
  switch (type) {
    case ACTIONS.SET_STACKS: {
      return {
        ...state,
        stacks: payload,
      };
    }
    case ACTIONS.SET_LAST_KEY: {
      return {
        ...state,
        lastKey: payload,
      };
    }
    case ACTIONS.SET_ORDER: {
      return {
        ...state,
        order: payload,
      };
    }
    case ACTIONS.SET_FOCUSED_STACK: {
      return {
        ...state,
        focusedStack: payload,
      };
    }
    case ACTIONS.SET_FOCUSED_ITEM: {
      return {
        ...state,
        focusedItem: payload,
      };
    }
    case ACTIONS.SET_STACK_INDEX: {
      return {
        ...state,
        stacksIndexes: {
          ...state.stacksIndexes,
          [payload.key]: payload.index,
        },
      };
    }
    case ACTIONS.SET_STACK_INDEXES: {
      return {
        ...state,
        stacksIndexes: payload,
      };
    }
    case ACTIONS.SET_STACKS_NAMES: {
      return {
        ...state,
        stacksNames: payload,
      };
    }
    default:
      return state;
  }
};
