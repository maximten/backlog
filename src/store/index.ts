import { createStore as createStoreInner, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import { lists } from 'src/store/lists';

const rootReduer = combineReducers({ lists });

export const createStore = () => createStoreInner(rootReduer,
  composeWithDevTools(applyMiddleware(thunk)));
