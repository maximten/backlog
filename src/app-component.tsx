import 'src/styles';

import React, { FC } from 'react';
import { Provider } from 'react-redux';
import { store } from 'src/store';
import { StackArrayComponent } from 'src/stack-array-component';
import { ShortcutsComponent } from 'src/shortcuts-component';
import { ContainerComponent } from './container-component';

export const AppComponent: FC = () => (
  <Provider store={store}>
    <ContainerComponent>
      <h1>Backlog</h1>
      <ShortcutsComponent />
      <StackArrayComponent />
    </ContainerComponent>
  </Provider>
);
