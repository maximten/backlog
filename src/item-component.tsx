import React, { FC, useCallback, useState, useEffect, useRef } from 'react';
import { StackItem, useStacks } from 'src/use-stacks';
import styled from 'styled-components';

type Props = {
  data: StackItem;
  index: number;
  stackKey: string;
  focusedStack: string;
  focusedItem: number | null;
}

const Form = styled.form`
    display: flex;
    flex-direction: column;
`;

export const ItemComponent: FC<Props> = ({
  data,
  index,
  stackKey: ownStackKey,
  focusedStack,
  focusedItem,
}) => {
  const { content } = data;
  const textareaRef = useRef(null);
  const [currentContent, setCurrentContent] = useState(content);

  const {
    modifyStackItem,
    setFocus,
  } = useStacks();

  const handleSubmit = useCallback((e) => {
    e.preventDefault();
    modifyStackItem(index, {
      ...data,
      content: e.target.value,
    });
  }, [currentContent, modifyStackItem]);

  const handleContentChange = useCallback((e) => {
    e.preventDefault();
    setCurrentContent(e.target.value);
    handleSubmit(e);
  }, [currentContent, modifyStackItem]);

  const handleFocus = useCallback((e) => {
    e.preventDefault();
    setFocus(ownStackKey, index);
  }, [index]);

  useEffect(() => {
    if (
      ownStackKey === focusedStack
      && index === focusedItem
    ) {
      textareaRef.current.focus();
    }
  }, [index, ownStackKey, focusedStack, focusedItem]);

  return (
    <Form onSubmit={handleSubmit}>
      <textarea
        ref={textareaRef}
        value={currentContent}
        onChange={handleContentChange}
        onFocus={handleFocus}
      />
    </Form>
  );
};
