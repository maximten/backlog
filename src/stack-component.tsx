import React, { FC } from 'react';
import { Stack } from 'src/use-stacks';
import { ItemComponent } from 'src/item-component';

type Props = {
  stack: Stack;
  stackKey: string;
  focusedStack: string;
  focusedItem: number | null;
}

export const StackComponent: FC<Props> = ({
  stack,
  stackKey,
  focusedStack,
  focusedItem,
}) => (
  <>
    {
      stack.map((item, key) => (
        <ItemComponent
          // TODO: add better key
          key={item.content}
          data={item}
          index={key}
          stackKey={stackKey}
          focusedStack={focusedStack}
          focusedItem={focusedItem}
        />
      ))
    }
  </>
);
