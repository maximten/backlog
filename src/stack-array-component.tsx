import React, { FC, useCallback, useEffect, useMemo } from 'react';
import { useStacks, createEmptyItem } from 'src/use-stacks';
import { useKeyboard, KeyCodes } from 'src/use-keyboard';
import { StackComponent } from 'src/stack-component';
import { useDispatch } from 'react-redux';
import { fetchState } from 'src/store/stacks';

type Props = {};

export const StackArrayComponent: FC<Props> = () => {
  const {
    stacks,
    pushStackItem,
    popStackItem,
    focusedStack,
    focusedItem,
    swapStackItemsUp,
    addStack,
    focusToLeftStack,
    focusToRightStack,
  } = useStacks();
  useKeyboard({
    [KeyCodes.SPACE]: [{
      isShiftPressed: true,
      callback: useCallback((e) => {
        e.preventDefault();
        pushStackItem(createEmptyItem());
      },
      [pushStackItem]),
    }, {
      isCtrlPresed: true,
      callback: useCallback((e) => {
        e.preventDefault();
        addStack();
      },
      [addStack]),
    }],
    [KeyCodes.BACKSPACE]: [{
      isShiftPressed: true,
      callback: useCallback((e) => {
        e.preventDefault();
        popStackItem();
      },
      [popStackItem]),
    }],
    [KeyCodes.UP]: [{
      isShiftPressed: true,
      callback: useCallback((e) => {
        e.preventDefault();
        swapStackItemsUp();
      },
      [swapStackItemsUp]),
    }],
    [KeyCodes.LEFT]: [{
      isShiftPressed: true,
      callback: useCallback((e) => {
        e.preventDefault();
        focusToLeftStack();
      },
      [focusToLeftStack]),
    }],
    [KeyCodes.RIGHT]: [{
      isShiftPressed: true,
      callback: useCallback((e) => {
        e.preventDefault();
        focusToRightStack();
      },
      [focusToRightStack]),
    }],
  });
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchState());
  }, [dispatch]);
  const stack = useMemo(() => stacks[focusedStack], [focusedStack, stacks]);
  return (
    <StackComponent
      stack={stack}
      stackKey={focusedStack}
      focusedStack={focusedStack}
      focusedItem={focusedItem}
    />
  );
};
