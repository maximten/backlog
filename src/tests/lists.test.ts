import { createStore } from 'src/store';
import { createItem,
  createList,
  createListsMap,
  parseListsMap,
  parseList,
  parseItem } from 'src/store/lists/parsers';

beforeAll(() => {
  const date = new Date(2020, 1, 1);
  let counter = 1;
  // eslint-disable-next-line no-extend-native
  Date.prototype.getTime = () => date.valueOf() + counter++;
});

describe('store', () => {
  test('lists should be in store', () => {
    const store = createStore();
    const { lists } = store.getState();
    expect(lists).toMatchSnapshot();
  });
});

describe('parsets', () => {
  test('parsers should be valid', () => {
    const item = createItem();
    const parsedItem = parseItem(item);
    expect(parsedItem).toEqual(item);
    const list = createList();
    const parsedList = parseList(list);
    expect(parsedList).toEqual(list);
    const listsMap = createListsMap();
    const parsedListsMap = parseListsMap(listsMap);
    expect(parsedListsMap).toEqual(listsMap);
  });
});
