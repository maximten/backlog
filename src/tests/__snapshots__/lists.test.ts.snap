// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store lists should be in store 1`] = `
Object {
  "focusedList": "list_1580504400001",
  "listsMap": Object {
    "list_1580504400001": Object {
      "focusedItem": "item_1580504400002",
      "itemSelection": 0,
      "items": Object {
        "item_1580504400002": Object {
          "data": "",
        },
      },
      "itemsOrder": Array [
        "item_1580504400002",
      ],
      "name": "",
    },
  },
  "listsOrder": Array [
    "list_1580504400001",
  ],
}
`;
